core/src/impl/Kokkos_ExecSpaceManager.hpp

```diff
@@ -83,6 +84,39 @@ using fence_t = std::enable_if_t<
     std::is_void_v<decltype(std::declval<T const&>().fence("name"))> &&
     std::is_void_v<decltype(T::impl_static_fence("name"))>>;
 
+template <class T>
+constexpr bool check_is_semiregular() {
+  static_assert(std::is_default_constructible_v<T>);
+  static_assert(std::is_copy_constructible_v<T>);
+  static_assert(std::is_move_constructible_v<T>);
+  static_assert(std::is_copy_assignable_v<T>);
+  static_assert(std::is_move_assignable_v<T>);
+  static_assert(std::is_destructible_v<T>);
+  return true;
+}
+
+template <class T>
+using equal_to_t =
+    decltype(std::declval<T const&>() == std::declval<T const&>());
+
+template <class T>
+using not_equal_to_t =
+    decltype(std::declval<T const&>() != std::declval<T const&>());
+
+template <class T>
+constexpr bool check_is_equality_comparable() {
+  using Kokkos::is_detected_exact_v;
+  static_assert(is_detected_exact_v<bool, equal_to_t, T>);
+  static_assert(is_detected_exact_v<bool, not_equal_to_t, T>);
+  return true;
+}
+
+template <class T>
+constexpr bool check_is_regular() {
+  static_assert(check_is_semiregular<T>() && check_is_equality_comparable<T>());
+  return true;
+}
+
```

Comment:
Would it make sense to define and `static_assert` all of this in a test or maybe something like `Kokkos_Core.cpp`, so that these declarations don't add compilation cost to every client code?

Reply:
This is a good question.  I was really careful when I wrote this and this header file is exclusively included in the backend source file that does the backend registration, meaning in no way visible when you include `<Kokkos_Core.hpp>` or any other public header.